include/xtensor/xhistogram.hpp

```diff
@@ -0,0 +1,307 @@
+/***************************************************************************
+* Copyright (c) 2016, Johan Mabille, Sylvain Corlay and Wolf Vollprecht    *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+/**
+ * @brief construct histogram
+ */
+
+#ifndef XTENSOR_HISTOGRAM_HPP
+#define XTENSOR_HISTOGRAM_HPP
+
+#include "xtensor.hpp"
+#include "xsort.hpp"
+
+namespace xt
+{
+
+    /**
+     * @ingroup basic_functions
+     * @brief Compute the histogram of a set of data.
+     *
+     * @param The data.
+     * @param The bin-edges.
+     * @param Weight factors corresponding to each data-point.
+     * @param If true the resulting integral is normalized to 1. [default: false]
+     * @return An one-dimensional xarray<double>, length: bin_edges.size()-1.
+     */
+    template<class E1, class E2, class E3>
+    inline auto histogram(
+        const E1 &data, const E2 &bin_edges, const E3 &weights, bool density=false
+    )
+    {
+        // counter type
+        using size_type = detail::common_size_type_t<E1,E2,E3>;
+
+        // basic checks
+        #ifdef XTENSOR_ENABLE_ASSERT
+
+            // - shape and size
+            XTENSOR_ASSERT( data     .dimension() == 1 );
+            XTENSOR_ASSERT( weights  .dimension() == 1 );
+            XTENSOR_ASSERT( bin_edges.dimension() == 1 );
+            XTENSOR_ASSERT( weights  .size() == data.size() );
+            XTENSOR_ASSERT( bin_edges.size() >= 2 );
+            // - bin-edges must be sorted
+            for (size_type i = 1 ; i < bin_edges.size() ; ++i)
+            {
+                XTENSOR_ASSERT( bin_edges[i] >= bin_edges[i - 1] );
+            }
+            // - data must be enclosed in the bins
+            XTENSOR_ASSERT( xt::amin(data)[0] >= bin_edges[0] );
+            XTENSOR_ASSERT( xt::amax(data)[0] <= bin_edges[bin_edges.size() - 1] );
+
+        #endif
+
+        // initialize output
+        xt::xtensor<double,1> count = xt::zeros<double>({bin_edges.size()-1});
```

Comment:
The `value_type` of the weights should be used instead of `double`

Reply:
@JohanMabille Not sure what you mean here: you refer to a line with `count` but you mention `weights`. 

In any case that `count` is `double` is a bit of an Achilles' heel. It has be floating point if `density=true`, but normally can be integer if `density=false`.

Reply:
@JohanMabille Fair enough. But I do need to type-cast thereafter to do `count[i] /= ( (bin_edges[i+1] - bin_edges[i]) * N );`. Or do you have a better solution?

Reply:
Many solutions here:

- state in the documentation that `weight` needs to hold floating point values when density is true. Simple, but no check at compile time
- make the computation in a temporary of type double, and then assign the double to count (with a cast). Note this can be compatible with the previous one
- Make a metafunctino for "computing" the type of `count`: if `weight` hold floating point values, use the same type for `count`, otherwise default to double (this + the first one is imo the better option)